File: ProjectPlan/generate_sprocs_from_schema.py
import re
import pathlib

base = pathlib.Path(__file__).parent
text = (base / 'Gayatri-Portal-DB-script.txt').read_text(encoding='utf-8')


def find_table_blocks(sql):
    blocks = []
    pos = 0
    while True:
        m = re.search(r'CREATE TABLE \[dbo\]\.\[([^\]]+)\]\s*\(', sql[pos:], re.I)
        if not m:
            break
        start = pos + m.start()
        name = m.group(1)
        paren_idx = start + m.end() - 1
        depth = 1
        i = paren_idx + 1
        while i < len(sql):
            ch = sql[i]
            if ch == '(':
                depth += 1
            elif ch == ')':
                depth -= 1
                if depth == 0:
                    break
            i += 1
        if i >= len(sql):
            break
        block_text = sql[start:i+1]
        blocks.append((name, block_text))
        pos = i + 1
    return blocks


def normalize_type(coltype):
    t = coltype.upper().strip()
    if t.startswith(('NVARCHAR', 'VARCHAR', 'NCHAR', 'CHAR', 'TEXT')):
        return 'NVARCHAR(MAX)' if 'MAX' in t else t
    if t.startswith(('DECIMAL', 'MONEY', 'FLOAT')):
        return t
    if t.startswith(('DATETIME2', 'DATETIME', 'DATE', 'TIME')):
        return t
    if t.startswith(('INT', 'SMALLINT', 'TINYINT', 'BIGINT')):
        return t
    if t.startswith('BIT'):
        return 'BIT'
    return t


def main():
    blocks = find_table_blocks(text)
    tables = []
    for name, block_text in blocks:
        body = block_text[block_text.index('(')+1: len(block_text)-1]
        cols = []
        for raw in body.splitlines():
            line = raw.strip()
            if not line or line.startswith('PRIMARY KEY') or line.startswith('CONSTRAINT') or line.startswith(')'):
                continue
            line = line.rstrip(',')
            m = re.match(r'\[([^\]]+)\]\s+([^\s]+(?:\([^\)]*\))?)(.*)$', line)
            if not m:
                continue
            colname, ctype, rest = m.groups()
            rest = rest.strip()
            dtype = ctype.upper()
            if not re.match(r'^(INT|BIGINT|SMALLINT|TINYINT|BIT|DECIMAL|MONEY|FLOAT|DATETIME2?|DATE|TIME|NVARCHAR|VARCHAR|NCHAR|CHAR|TEXT)', dtype):
                continue
            cols.append({
                'colname': colname,
                'ctype': ctype,
                'is_identity': 'IDENTITY' in rest,
                'is_nullable': 'NOT NULL' not in rest,
            })
        pk = None
        pkm = re.search(r'PRIMARY KEY CLUSTERED \([^\)]*\[([^\]]+)\] ASC', block_text)
        if pkm:
            pk = pkm.group(1)
        else:
            for c in cols:
                if c['is_identity']:
                    pk = c['colname']
                    break
            if not pk and cols:
                pk = cols[0]['colname']
        if pk:
            tables.append((name, pk, cols))

    out = []
    out.append('USE [GayatriCateringPortal];\nGO\n\nSET ANSI_NULLS ON\nGO\nSET QUOTED_IDENTIFIER ON\nGO\n')
    for name, pk, cols in tables:
        non_id_cols = [c for c in cols if not c['is_identity']]
        params = []
        for c in non_id_cols:
            ptype = normalize_type(c['ctype'])
            default = ' = NULL' if c['is_nullable'] else ''
            if ptype.upper() == 'BIT' and c['is_nullable']:
                default = ' = NULL'
            params.append(f"    @{c['colname']} {ptype}{default}")

        out.append(f'-- Procedures for {name}')
        out.append(f"IF OBJECT_ID(N'SP_Create{name}', N'P') IS NOT NULL DROP PROCEDURE SP_Create{name};")
        out.append('GO')
        out.append(f'CREATE PROCEDURE SP_Create{name}')
        if params:
            out.append('(' + ',\n'.join(params) + ')')
        out.append('AS')
        out.append('BEGIN')
        out.append('    SET NOCOUNT ON;')
        if params:
            cols_list = ', '.join(f'[{c["colname"]}]' for c in non_id_cols)
            vals_list = ', '.join(f'@{c["colname"]}' for c in non_id_cols)
            if any(c['colname'].lower() == 'createddate' for c in non_id_cols):
                vals_list = vals_list.replace('@CreatedDate', 'COALESCE(@CreatedDate, GETDATE())')
            out.append(f'    INSERT INTO [dbo].[{name}] ({cols_list})')
            out.append(f'    VALUES ({vals_list});')
            out.append('    SELECT SCOPE_IDENTITY() AS NewId;')
        else:
            out.append(f'    INSERT INTO [dbo].[{name}] DEFAULT VALUES;')
            out.append('    SELECT SCOPE_IDENTITY() AS NewId;')
        out.append('END')
        out.append('GO')

        out.append(f"IF OBJECT_ID(N'SP_Update{name}', N'P') IS NOT NULL DROP PROCEDURE SP_Update{name};")
        out.append('GO')
        upd_params = [f"    @{pk} {normalize_type(next(c['ctype'] for c in cols if c['colname'] == pk))}"]
        upd_params += params
        out.append(f'CREATE PROCEDURE SP_Update{name}')
        out.append('(' + ',\n'.join(upd_params) + ')')
        out.append('AS')
        out.append('BEGIN')
        out.append('    SET NOCOUNT ON;')
        if params:
            set_list = ',\n'.join(f'    [{c["colname"]}] = @{c["colname"]}' for c in non_id_cols)
            out.append(f'    UPDATE [dbo].[{name}]')
            out.append('    SET ' + set_list)
            out.append(f'    WHERE [{pk}] = @{pk};')
        out.append('    SELECT @@ROWCOUNT AS RowsAffected;')
        out.append('END')
        out.append('GO')

        out.append(f"IF OBJECT_ID(N'SP_Get{name}ById', N'P') IS NOT NULL DROP PROCEDURE SP_Get{name}ById;")
        out.append('GO')
        out.append(f'CREATE PROCEDURE SP_Get{name}ById')
        out.append('(')
        out.append(f'    @{pk} {normalize_type(next(c["ctype"] for c in cols if c["colname"] == pk))}')
        out.append(')')
        out.append('AS')
        out.append('BEGIN')
        out.append('    SET NOCOUNT ON;')
        out.append(f'    SELECT * FROM [dbo].[{name}] WHERE [{pk}] = @{pk};')
        out.append('END')
        out.append('GO')

        out.append(f"IF OBJECT_ID(N'Get{name}', N'P') IS NOT NULL DROP PROCEDURE Get{name};")
        out.append('GO')
        out.append(f'CREATE PROCEDURE Get{name}')
        out.append('AS')
        out.append('BEGIN')
        out.append('    SET NOCOUNT ON;')
        where_deleted = any(c['colname'].lower() == 'isdeleted' for c in cols)
        out.append(f'    SELECT * FROM [dbo].[{name}]' + (' WHERE [IsDeleted] = 0' if where_deleted else '') + ';')
        out.append('END')
        out.append('GO')

        out.append(f"IF OBJECT_ID(N'Delete{name}ById', N'P') IS NOT NULL DROP PROCEDURE Delete{name}ById;")
        out.append('GO')
        out.append(f'CREATE PROCEDURE Delete{name}ById')
        out.append('(')
        out.append(f'    @{pk} {normalize_type(next(c["ctype"] for c in cols if c["colname"] == pk))}')
        out.append(')')
        out.append('AS')
        out.append('BEGIN')
        out.append('    SET NOCOUNT ON;')
        if any(c['colname'].lower() == 'isdeleted' for c in cols):
            out.append(f'    UPDATE [dbo].[{name}] SET [IsDeleted] = 1 WHERE [{pk}] = @{pk};')
        else:
            out.append(f'    DELETE FROM [dbo].[{name}] WHERE [{pk}] = @{pk};')
        out.append('    SELECT @@ROWCOUNT AS RowsAffected;')
        out.append('END')
        out.append('GO')

        out.append(f"IF OBJECT_ID(N'ActiveInActive{name}ById', N'P') IS NOT NULL DROP PROCEDURE ActiveInActive{name}ById;")
        out.append('GO')
        out.append(f'CREATE PROCEDURE ActiveInActive{name}ById')
        out.append('(')
        out.append(f'    @{pk} {normalize_type(next(c["ctype"] for c in cols if c["colname"] == pk))},')
        out.append('    @IsActive BIT')
        out.append(')')
        out.append('AS')
        out.append('BEGIN')
        out.append('    SET NOCOUNT ON;')
        if any(c['colname'].lower() == 'isactive' for c in cols):
            out.append(f'    UPDATE [dbo].[{name}] SET [IsActive] = @IsActive WHERE [{pk}] = @{pk};')
            out.append('    SELECT @@ROWCOUNT AS RowsAffected;')
        else:
            out.append(f'    RAISERROR(\'Table [{name}] does not contain an IsActive column.\', 16, 1);')
            out.append('    RETURN;')
        out.append('END')
        out.append('GO')

    output_path = base / 'Gayatri-Portal-DB-sprocs.sql'
    output_path.write_text('\n'.join(out), encoding='utf-8')
    print(f'Wrote {output_path} with {len(tables)} tables.')


if __name__ == '__main__':
    main()

File: ProjectPlan/generate_sprocs_from_schema.js
const fs = require('fs');
const path = require('path');
const base = __dirname;
const text = fs.readFileSync(path.join(base, 'Gayatri-Portal-DB-script.txt'), 'utf8');

function findTableBlocks(sql) {
  const blocks = [];
  const createRegex = /^CREATE TABLE \[dbo\]\.\[([^\]]+)\]\s*\(/i;
  const lines = sql.split(/\r?\n/);
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    const m = line.match(createRegex);
    if (!m) continue;

    const name = m[1];
    const blockLines = [lines[i]];
    let depth = 0;
    let foundStart = false;

    for (let j = i; j < lines.length; j++) {
      const currentLine = lines[j];
      const openCount = (currentLine.match(/\(/g) || []).length;
      const closeCount = (currentLine.match(/\)/g) || []).length;

      if (!foundStart) {
        if (currentLine.includes('(')) {
          foundStart = true;
          depth += openCount;
          depth -= closeCount;
        }
      } else {
        depth += openCount;
        depth -= closeCount;
      }

      blockLines.push(currentLine);
      if (foundStart && depth <= 0) {
        i = j;
        break;
      }
    }

    blocks.push([name, blockLines.join('\n')]);
  }
  return blocks;
}

function normalizeType(coltype) {
  const t = coltype.toUpperCase().trim();
  if (t.startsWith('NVARCHAR') || t.startsWith('VARCHAR') || t.startsWith('NCHAR') || t.startsWith('CHAR') || t.startsWith('TEXT')) {
    return t.includes('MAX') ? 'NVARCHAR(MAX)' : t;
  }
  if (t.startsWith('DECIMAL') || t.startsWith('MONEY') || t.startsWith('FLOAT')) return t;
  if (t.startsWith('DATETIME2') || t.startsWith('DATETIME') || t.startsWith('DATE') || t.startsWith('TIME')) return t;
  if (t.startsWith('INT') || t.startsWith('SMALLINT') || t.startsWith('TINYINT') || t.startsWith('BIGINT')) return t;
  if (t.startsWith('BIT')) return 'BIT';
  return t;
}

const blocks = findTableBlocks(text);
const tables = [];
for (const [name, blockText] of blocks) {
  const firstParen = blockText.indexOf('(');
  const lastParen = blockText.lastIndexOf(')');
  const body = firstParen >= 0 && lastParen > firstParen ? blockText.slice(firstParen + 1, lastParen) : '';
  const cols = [];
  for (const raw of body.split(/\r?\n/)) {
    const line = raw.trim();
    if (!line || line.startsWith('PRIMARY KEY') || line.startsWith('CONSTRAINT') || line.startsWith(')')) continue;
    const m = line.replace(/,$/, '').match(/\[([^\]]+)\]\s+([^\s]+(?:\([^\)]*\))?)(.*)$/i);
    if (!m) {
      const fallback = line.match(/^([A-Za-z_][A-Za-z0-9_]*)\s+([^\s]+(?:\([^\)]*\))?)(.*)$/);
      if (!fallback) continue;
      const colname = fallback[1].replace(/^\[/, '').replace(/\]$/, '');
      const ctype = fallback[2].replace(/^\[/, '').replace(/\]$/, '');
      const rest = fallback[3].trim();
      const dtype = ctype.toUpperCase();
      if (!/^(INT|BIGINT|SMALLINT|TINYINT|BIT|DECIMAL|MONEY|FLOAT|DATETIME2?|DATE|TIME|NVARCHAR|VARCHAR|NCHAR|CHAR|TEXT)/.test(dtype)) continue;
      cols.push({ colname, ctype, isIdentity: rest.includes('IDENTITY'), isNullable: !rest.includes('NOT NULL') });
      continue;
    }
    const colname = m[1];
    const ctype = m[2].replace(/^\[/, '').replace(/\]$/, '');
    const rest = m[3].trim();
    const dtype = ctype.toUpperCase();
    if (!/^(INT|BIGINT|SMALLINT|TINYINT|BIT|DECIMAL|MONEY|FLOAT|DATETIME2?|DATE|TIME|NVARCHAR|VARCHAR|NCHAR|CHAR|TEXT)/.test(dtype)) continue;
    cols.push({ colname, ctype, isIdentity: rest.includes('IDENTITY'), isNullable: !rest.includes('NOT NULL') });
  }
  let pk = null;
  const pkm = blockText.match(/PRIMARY KEY CLUSTERED \([^\)]*\[([^\]]+)\] ASC/);
  if (pkm) pk = pkm[1];
  else {
    const identityCol = cols.find(c => c.isIdentity);
    pk = identityCol ? identityCol.colname : (cols[0] ? cols[0].colname : null);
  }
  if (pk) tables.push({ name, pk, cols });
}

const out = [];
out.push('USE [GayatriCateringPortal];\nGO\n\nSET ANSI_NULLS ON\nGO\nSET QUOTED_IDENTIFIER ON\nGO\n');
for (const { name, pk, cols } of tables) {
  const nonIdCols = cols.filter(c => !c.isIdentity);
  const params = nonIdCols.map(c => {
    const ptype = normalizeType(c.ctype);
    const defaultVal = c.isNullable ? ' = NULL' : '';
    return `    @${c.colname} ${ptype}${defaultVal}`;
  });

  out.push(`-- Procedures for ${name}`);
  out.push(`IF OBJECT_ID(N'SP_Create${name}', N'P') IS NOT NULL DROP PROCEDURE SP_Create${name};`);
  out.push('GO');
  out.push(`CREATE PROCEDURE SP_Create${name}`);
  if (params.length) out.push(`(${params.join(',\n')})`);
  out.push('AS');
  out.push('BEGIN');
  out.push('    SET NOCOUNT ON;');
  if (params.length) {
    const colsList = nonIdCols.map(c => `[${c.colname}]`).join(', ');
    let valsList = nonIdCols.map(c => `@${c.colname}`).join(', ');
    if (nonIdCols.some(c => c.colname.toLowerCase() === 'createddate')) {
      valsList = valsList.replace('@CreatedDate', 'COALESCE(@CreatedDate, GETDATE())');
    }
    out.push(`    INSERT INTO [dbo].[${name}] (${colsList})`);
    out.push(`    VALUES (${valsList});`);
    out.push('    SELECT SCOPE_IDENTITY() AS NewId;');
  } else {
    out.push(`    INSERT INTO [dbo].[${name}] DEFAULT VALUES;`);
    out.push('    SELECT SCOPE_IDENTITY() AS NewId;');
  }
  out.push('END');
  out.push('GO');

  out.push(`IF OBJECT_ID(N'SP_Update${name}', N'P') IS NOT NULL DROP PROCEDURE SP_Update${name};`);
  out.push('GO');
  const updParams = [`    @${pk} ${normalizeType(cols.find(c => c.colname === pk).ctype)}`, ...params];
  out.push(`CREATE PROCEDURE SP_Update${name}`);
  out.push(`(${updParams.join(',\n')})`);
  out.push('AS');
  out.push('BEGIN');
  out.push('    SET NOCOUNT ON;');
  if (params.length) {
    const setList = nonIdCols.map(c => `    [${c.colname}] = @${c.colname}`).join('\n');
    out.push(`    UPDATE [dbo].[${name}]`);
    out.push('    SET ' + setList);
    out.push(`    WHERE [${pk}] = @${pk};`);
  }
  out.push('    SELECT @@ROWCOUNT AS RowsAffected;');
  out.push('END');
  out.push('GO');

  out.push(`IF OBJECT_ID(N'SP_Get${name}ById', N'P') IS NOT NULL DROP PROCEDURE SP_Get${name}ById;`);
  out.push('GO');
  out.push(`CREATE PROCEDURE SP_Get${name}ById`);
  out.push('(');
  out.push(`    @${pk} ${normalizeType(cols.find(c => c.colname === pk).ctype)}`);
  out.push(')');
  out.push('AS');
  out.push('BEGIN');
  out.push('    SET NOCOUNT ON;');
  out.push(`    SELECT * FROM [dbo].[${name}] WHERE [${pk}] = @${pk};`);
  out.push('END');
  out.push('GO');

  out.push(`IF OBJECT_ID(N'Get${name}', N'P') IS NOT NULL DROP PROCEDURE Get${name};`);
  out.push('GO');
  out.push(`CREATE PROCEDURE Get${name}`);
  out.push('AS');
  out.push('BEGIN');
  out.push('    SET NOCOUNT ON;');
  const whereDeleted = cols.some(c => c.colname.toLowerCase() === 'isdeleted');
  out.push(`    SELECT * FROM [dbo].[${name}]${whereDeleted ? ' WHERE [IsDeleted] = 0' : ''};`);
  out.push('END');
  out.push('GO');

  out.push(`IF OBJECT_ID(N'Delete${name}ById', N'P') IS NOT NULL DROP PROCEDURE Delete${name}ById;`);
  out.push('GO');
  out.push(`CREATE PROCEDURE Delete${name}ById`);
  out.push('(');
  out.push(`    @${pk} ${normalizeType(cols.find(c => c.colname === pk).ctype)}`);
  out.push(')');
  out.push('AS');
  out.push('BEGIN');
  out.push('    SET NOCOUNT ON;');
  if (whereDeleted) {
    out.push(`    UPDATE [dbo].[${name}] SET [IsDeleted] = 1 WHERE [${pk}] = @${pk};`);
  } else {
    out.push(`    DELETE FROM [dbo].[${name}] WHERE [${pk}] = @${pk};`);
  }
  out.push('    SELECT @@ROWCOUNT AS RowsAffected;');
  out.push('END');
  out.push('GO');

  out.push(`IF OBJECT_ID(N'ActiveInActive${name}ById', N'P') IS NOT NULL DROP PROCEDURE ActiveInActive${name}ById;`);
  out.push('GO');
  out.push(`CREATE PROCEDURE ActiveInActive${name}ById`);
  out.push('(');
  out.push(`    @${pk} ${normalizeType(cols.find(c => c.colname === pk).ctype)},`);
  out.push('    @IsActive BIT');
  out.push(')');
  out.push('AS');
  out.push('BEGIN');
  out.push('    SET NOCOUNT ON;');
  if (cols.some(c => c.colname.toLowerCase() === 'isactive')) {
    out.push(`    UPDATE [dbo].[${name}] SET [IsActive] = @IsActive WHERE [${pk}] = @${pk};`);
    out.push('    SELECT @@ROWCOUNT AS RowsAffected;');
  } else {
    out.push(`    RAISERROR('Table [${name}] does not contain an IsActive column.', 16, 1);`);
    out.push('    RETURN;');
  }
  out.push('END');
  out.push('GO');
}
fs.writeFileSync(path.join(base, 'Gayatri-Portal-DB-sprocs.sql'), out.join('\n'), 'utf8');
console.log('Wrote', path.join(base, 'Gayatri-Portal-DB-sprocs.sql'), 'for', tables.length, 'tables.');
